arches/app/views/api/controlled_lists.py

```diff
@@ -0,0 +1,413 @@
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ControlledListItemView(View):
+    def post(self, request, **kwargs):
+        data = JSONDeserializer().deserialize(request.body)
+
+        # Update
+        if "id" in kwargs:
+            raise NotImplementedError
+
+        # Insert
+        parent_id = data["parent_id"]
+        controlled_list_id = data["controlled_list_id"]
+
+        try:
+            with transaction.atomic():
+                controlled_list = (
+                    ControlledList.objects.filter(pk=controlled_list_id)
+                    .annotate(
+                        max_sortorder=Max(
+                            "controlled_list_items__sortorder", default=-1
+                        )
+                    )
+                    .get()
+                )
+                item = ControlledListItem.objects.create(
+                    controlled_list=controlled_list,
+                    sortorder=controlled_list.max_sortorder + 1,
+                    parent_id=parent_id,
+                )
+        except ControlledList.DoesNotExist:
+            return JSONErrorResponse(status=HTTPStatus.NOT_FOUND)
+
+        return JSONResponse(item.serialize(), status=HTTPStatus.CREATED)
+
+    def patch(self, request, **kwargs):
+        item_id: UUID = kwargs.get("id")
+        data = JSONDeserializer().deserialize(request.body)
+        item = ControlledListItem(id=item_id, **data)
+
+        update_fields = list(data)
+        if not update_fields:
+            return JSONErrorResponse(status=HTTPStatus.BAD_REQUEST)
+        exclude_fields = {f for f in field_names(item) if f not in update_fields}
+        try:
+            item._state.adding = False
+            item.full_clean(exclude=exclude_fields)
+            with transaction.atomic():
+                item.save(update_fields=update_fields)
+                if "parent_id" in update_fields:
+                    # Check for recursive structure
+                    unused = item.parent.serialize()
+        except RecursionError:
+            return JSONErrorResponse(
+                message=_("Recursive structure detected."),
+                status=HTTPStatus.BAD_REQUEST,
+            )
+        except ValidationError as ve:
+            return JSONErrorResponse(
+                message="\n".join(ve.messages), status=HTTPStatus.BAD_REQUEST
+            )
+
+        return JSONResponse(status=HTTPStatus.NO_CONTENT)
+
+    def delete(self, request, **kwargs):
+        item_id: UUID = kwargs.get("id")
+        objs_deleted, unused = ControlledListItem.objects.filter(pk=item_id).delete()
+        if not objs_deleted:
+            return JSONErrorResponse(status=HTTPStatus.NOT_FOUND)
+        return JSONResponse(status=HTTPStatus.NO_CONTENT)
+
+
+@method_decorator(
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ControlledListItemValueView(View):
+    def post(self, request, **kwargs):
```

Comment:
IMO this is a pretty hefty/complex view that's fairly hard to read. And I don't believe the answer is how it was previously, eg building accessory methods in the view. Why not make a service class or util or something?

spitballing: 

```
from django.core.exceptions import ValidationError
from uuid import UUID

class ControlledListItemValueService:
    @staticmethod
    def create_value(data):
        value = ControlledListItemValue(
            controlled_list_item_id=UUID(data["item_id"]),
            valuetype_id=data["valuetype_id"],
            language_id=data["language_id"],
            value=data["value"],
        )
        ControlledListItemValueService.validate(value)
        value.save()
        return value

    @staticmethod
    def update_value(value_id, data):
        value = ControlledListItemValueService.get_value(value_id)
        if not value:
            return None

        value.value = data["value"]
        value.valuetype_id = data["valuetype_id"]
        value.language = ControlledListItemValueService.get_language(data["language_id"])
        ControlledListItemValueService.validate(value)
        value.save()
        return value

    @staticmethod
    def validate(value):
        try:
            value.full_clean()
        except ValidationError as ve:
            raise ve

    @staticmethod
    def get_value(value_id):
        try:
            return ControlledListItemValue.objects.values_without_images().get(pk=value_id)
        except ControlledListItemValue.DoesNotExist:
            return None

    @staticmethod
    def get_language(language_id):
        try:
            return Language.objects.get(code=language_id)
        except Language.DoesNotExist:
            return None
  ```

So the view looks like:

```
class ControlledListItemValueView(View):
    def post(self, request, **kwargs):
        data = JSONDeserializer().deserialize(request.body)
        value_id = kwargs.get("id")

        try:
            if not value_id:
                value = ControlledListItemValueService.create_value(data)
                return JSONResponse(value.serialize(), status=HTTPStatus.CREATED)
            else:
                value = ControlledListItemValueService.update_value(value_id, data)
                if not value:
                    return JSONErrorResponse(status=HTTPStatus.NOT_FOUND)
                return JSONResponse(value.serialize())
        except ValidationError as ve:
            return JSONErrorResponse(
                message="\n".join(ve.messages), status=HTTPStatus.BAD_REQUEST
            )
        except Language.DoesNotExist:
            return JSONErrorResponse(status=HTTPStatus.NOT_FOUND
            ```

Reply:
Thanks for fleshing this out; I’m glad we’re innovating on this and considering alternatives.

I’m down with the motivation to keep slicing the view thinner, but in terms of the implementation, I'll just echo what I said [here](https://github.com/archesproject/arches/pull/10569#discussion_r1646550202) and [here](https://github.com/archesproject/arches/pull/10569#discussion_r1657250707), which is that I think we need have a thoughtful examination of where to put this proposed validation layer. (This service layer you sketched strikes me as mostly a _deserialize 'n validate_ layer, which is all good, because if it were anything else, I'd probably fall back on James Bennett's "[Against Service Layers in Django](https://www.b-list.org/weblog/2020/mar/16/no-service/)" to make the case against. 😉)

So, if you agree with me that we're talking about a _deserialize-'n-validate_ layer, my next question is, why not use DRF, which is more standard? I think we need to be using framework tools where possible.

Then if the question is _why didn't I use DRF from the start_, the answer is:
- it seemed lower priority than the other fish we were frying
- it might require more consensus-building than we had time for
- it seemed of low benefit for such simple models
- I was not certain how much customization would be needed for fields that are not part of the model, like the dynamic calculation of children
- We have existing permissions layers I expected to be reusing

Then, if the question is, _can we do something like DRF_, in looking at the sketch, which I know is just a sketch, I have some concerns:
- I don't love one-line hooks that do one thing (or one thing and immediately reraise), which I guess is why I don't see this as a win -- we replace one inline call to a call to a hook that itself does one thing?
- Or two things if you add save(), but I'm pretty firmly against mixing validate() and save(). Validate should happen at the user input boundary, and save should happen at the database boundary, and the only reason to suppose those are happening in quick succession is in a view: it seems like memorializing the wrong abstraction to marry them off and put them someplace else. (It needs to be easy to reason about where the user input boundaries are, so you don't validate too much or not enough. The most obvious place is the very first line of a view.)

Before we start the vue cutover in core we definitely need to continue this conversation.

***
That said, there are a couple improvements this has inspired me to make:
- in a couple places I'm doing `Model(**data)` where there were no fields to rename or no concern about overposting. I can do this in a couple more places to reduce some verbosity (and frankly improve correctness, I think missing parameters would have raised unhandled KeyErrors as is 🤕 )
- A couple places return 404 for a bad language code, but I can probably just let the ValidationError from full_clean() take care of that with a more appropriate 400
***
Let me know what you think.

Reply:
I'm so down for that. Will update 👍